File: slr1.ml
(** SLR(1) bottom-up parser *)

open Utils
open Grammar
open FirstFollow

(** LR(0) item: (production, dot_position) *)
type item = production * int

(** SLR(1) action *)
type action =
  | Shift of int
  | Reduce of production
  | Accept
  | Error

(** Item set (state) *)
module ItemSet = Set.Make(struct
  type t = item
  let compare (p1, pos1) (p2, pos2) =
    let c = compare_symbol p1.lhs p2.lhs in
    if c <> 0 then c else
    let c = List.compare compare_symbol p1.rhs p2.rhs in
    if c <> 0 then c else
    Int.compare pos1 pos2
end)

(** SLR(1) parser type *)
type t = {
  grammar : Grammar.t;
  augmented_start : symbol;  (* S' for augmented grammar *)
  states : ItemSet.t array;
  action_table : (int * symbol, action) Hashtbl.t;
  goto_table : (int * symbol, int) Hashtbl.t;
  first_sets : first_sets;
  follow_sets : follow_sets;
}

(** Exception raised when grammar is not SLR(1) *)
exception Not_SLR1 of string

(** Compare items *)
let compare_items (p1, pos1) (p2, pos2) =
  let c = compare_symbol p1.lhs p2.lhs in
  if c <> 0 then c else
  let c = List.compare compare_symbol p1.rhs p2.rhs in
  if c <> 0 then c else
  Int.compare pos1 pos2

(** Convert item to string *)
let item_to_string (prod, dot_pos) =
  let lhs_str = symbol_to_string prod.lhs in
  let rhs = prod.rhs in
  let rec build_rhs pos acc = function
    | [] ->
        if pos = dot_pos then
          List.rev ("•" :: acc)
        else
          List.rev acc
    | sym :: rest ->
        if pos = dot_pos then
          build_rhs (pos + 1) (symbol_to_string sym :: "•" :: acc) rest
        else
          build_rhs (pos + 1) (symbol_to_string sym :: acc) rest
  in
  let rhs_str = String.concat " " (build_rhs 0 [] rhs) in
  lhs_str ^ " → " ^ rhs_str

(** Get symbol after dot in an item, if any *)
let symbol_after_dot (prod, dot_pos) =
  if dot_pos < List.length prod.rhs then
    Some (List.nth prod.rhs dot_pos)
  else
    None

(** Compute closure of a set of items
    For each item [A → α•Bβ] where B is nonterminal,
    add all items [B → •γ] for each production B → γ *)
let closure grammar items =
  let rec aux current =
    let new_items = ItemSet.fold (fun item acc ->
      match symbol_after_dot item with
      | Some sym when is_nonterminal sym ->
          let prods = get_productions grammar sym in
          let new_items = List.map (fun prod -> (prod, 0)) prods in
          List.fold_left (fun s i -> ItemSet.add i s) acc new_items
      | _ -> acc
    ) current current in

    if ItemSet.equal current new_items then
      current
    else
      aux new_items
  in
  aux items

(** Compute goto(I, X) - the set of items obtained by moving dot over X *)
let goto grammar items symbol =
  let moved = ItemSet.fold (fun (prod, dot_pos) acc ->
    match symbol_after_dot (prod, dot_pos) with
    | Some sym when compare_symbol sym symbol = 0 ->
        ItemSet.add (prod, dot_pos + 1) acc
    | _ -> acc
  ) items ItemSet.empty in
  closure grammar moved

(** Build the canonical LR(0) collection of item sets *)
let build_lr0_automaton grammar start_prod =
  let initial_item = (start_prod, 0) in
  let initial_state = closure grammar (ItemSet.singleton initial_item) in

  let states = ref [initial_state] in
  let state_map = Hashtbl.create 100 in
  Hashtbl.add state_map initial_state 0;

  let transitions = Hashtbl.create 100 in

  let rec process_states worklist =
    match worklist with
    | [] -> ()
    | state :: rest ->
        let state_id = Hashtbl.find state_map state in

        (* Get all symbols that can be shifted *)
        let symbols = ItemSet.fold (fun item acc ->
          match symbol_after_dot item with
          | Some sym -> SymbolSet.add sym acc
          | None -> acc
        ) state SymbolSet.empty in

        (* For each symbol, compute goto and add new states *)
        let new_worklist = SymbolSet.fold (fun sym worklist_acc ->
          let next_state = goto grammar state sym in
          if not (ItemSet.is_empty next_state) then begin
            if not (Hashtbl.mem state_map next_state) then begin
              let new_id = List.length !states in
              states := !states @ [next_state];
              Hashtbl.add state_map next_state new_id;
              Hashtbl.add transitions (state_id, sym) new_id;
              next_state :: worklist_acc
            end else begin
              let next_id = Hashtbl.find state_map next_state in
              Hashtbl.add transitions (state_id, sym) next_id;
              worklist_acc
            end
          end else
            worklist_acc
        ) symbols rest in

        process_states new_worklist
  in

  process_states [initial_state];
  (Array.of_list !states, transitions)

(** Build ACTION and GOTO tables for SLR(1) *)
let build_tables grammar states transitions first_sets follow_sets augmented_start start_prod =
  let action_table = Hashtbl.create 200 in
  let goto_table = Hashtbl.create 200 in

  (* Build state_map for reverse lookup *)
  let state_map = Hashtbl.create (Array.length states) in
  Array.iteri (fun i state ->
    Hashtbl.add state_map state i
  ) states;

  (* Process each state *)
  Array.iteri (fun state_id state ->
    (* Process each item in the state *)
    ItemSet.iter (fun (prod, dot_pos) ->
      if dot_pos < List.length prod.rhs then begin
        (* Shift items: [A → α•aβ] where a is terminal *)
        match symbol_after_dot (prod, dot_pos) with
        | Some sym when is_terminal sym ->
            let key = (state_id, sym) in
            if Hashtbl.mem transitions key then begin
              let next_state = Hashtbl.find transitions key in
              let action_key = (state_id, sym) in
              if Hashtbl.mem action_table action_key then
                raise (Not_SLR1 (Printf.sprintf
                  "Shift/Shift or Shift/Reduce conflict at state %d, symbol %s"
                  state_id (symbol_to_string sym)))
              else
                Hashtbl.add action_table action_key (Shift next_state)
            end
        | _ -> ()
      end else begin
        (* Reduce items: [A → α•] *)
        if prod.lhs = augmented_start then begin
          (* Accept item: [S' → S•] *)
          let action_key = (state_id, EndMarker) in
          Hashtbl.add action_table action_key Accept
        end else begin
          (* Reduce on FOLLOW(A) *)
          let follow_a = follow_of_symbol follow_sets prod.lhs in
          SymbolSet.iter (fun sym ->
            let action_key = (state_id, sym) in
            if Hashtbl.mem action_table action_key then begin
              let existing = Hashtbl.find action_table action_key in
              match existing with
              | Shift _ ->
                  raise (Not_SLR1 (Printf.sprintf
                    "Shift/Reduce conflict at state %d, symbol %s"
                    state_id (symbol_to_string sym)))
              | Reduce other_prod ->
                  raise (Not_SLR1 (Printf.sprintf
                    "Reduce/Reduce conflict at state %d, symbol %s:\n  %s\n  %s"
                    state_id (symbol_to_string sym)
                    (production_to_string other_prod)
                    (production_to_string prod)))
              | _ -> ()
            end else
              Hashtbl.add action_table action_key (Reduce prod)
          ) follow_a
        end
      end
    ) state;

    (* Build GOTO table for nonterminals *)
    Hashtbl.iter (fun (src, sym) dst ->
      if src = state_id && is_nonterminal sym then
        Hashtbl.add goto_table (state_id, sym) dst
    ) transitions
  ) states;

  (action_table, goto_table)

(** Build SLR(1) parser *)
let build grammar first_sets follow_sets =
  (* Create augmented grammar with S' → S *)
  let start = get_start_symbol grammar in
  let augmented_start = Nonterminal '\'' in  (* S' represented as ' *)
  let start_prod = { lhs = augmented_start; rhs = [start] } in

  (* Build LR(0) automaton *)
  let (states, transitions) = build_lr0_automaton grammar start_prod in

  (* Build ACTION and GOTO tables *)
  let (action_table, goto_table) =
    build_tables grammar states transitions first_sets follow_sets augmented_start start_prod in

  {
    grammar;
    augmented_start;
    states;
    action_table;
    goto_table;
    first_sets;
    follow_sets;
  }

(** Parse a string using SLR(1) shift-reduce algorithm *)
let parse parser input_str =
  (* Convert input to symbols and add $ *)
  let input_chars = List.init (String.length input_str) (String.get input_str) in
  let input_symbols = List.map char_to_symbol input_chars in
  let input = input_symbols @ [EndMarker] in

  (* Initialize stack with state 0 *)
  let initial_stack = [0] in

  (* Parsing loop *)
  let rec parse_loop stack symbols input =
    match input with
    | [] -> false  (* Unexpected end *)
    | curr_sym :: input_rest ->
        let state = List.hd stack in
        let action_key = (state, curr_sym) in

        let action =
          try Hashtbl.find parser.action_table action_key
          with Not_found -> Error
        in

        match action with
        | Accept -> true
        | Error -> false
        | Shift next_state ->
            (* Push symbol and next state *)
            parse_loop (next_state :: stack) (curr_sym :: symbols) input_rest
        | Reduce prod ->
            (* Pop |rhs| symbols and states *)
            let rhs_len = List.length prod.rhs in
            let rhs_len = if prod.rhs = [Epsilon] then 0 else rhs_len in
            let stack' = List.filteri (fun i _ -> i >= rhs_len) stack in
            let symbols' = List.filteri (fun i _ -> i >= rhs_len) symbols in

            (* Get state at top of stack after popping *)
            let state' = List.hd stack' in

            (* Find next state via GOTO *)
            let goto_key = (state', prod.lhs) in
            begin
              try
                let next_state = Hashtbl.find parser.goto_table goto_key in
                parse_loop (next_state :: stack') (prod.lhs :: symbols') input
              with Not_found -> false
            end
  in

  try
    parse_loop initial_stack [] input
  with _ -> false

(** Print the LR(0) automaton states *)
let print_states parser =
  print_endline "\nLR(0) Automaton States:";
  Array.iteri (fun i state ->
    Printf.printf "\nState %d:\n" i;
    ItemSet.iter (fun item ->
      Printf.printf "  %s\n" (item_to_string item)
    ) state
  ) parser.states

(** Print ACTION and GOTO tables *)
let print_tables parser =
  print_endline "\nACTION Table:";
  let action_entries = Hashtbl.fold (fun (state, sym) action acc ->
    (state, sym, action) :: acc
  ) parser.action_table [] in
  let sorted_actions = List.sort (fun (s1, sym1, _) (s2, sym2, _) ->
    let c = Int.compare s1 s2 in
    if c = 0 then compare_symbol sym1 sym2 else c
  ) action_entries in
  List.iter (fun (state, sym, action) ->
    let action_str = match action with
      | Shift s -> "shift " ^ string_of_int s
      | Reduce p -> "reduce " ^ production_to_string p
      | Accept -> "accept"
      | Error -> "error"
    in
    Printf.printf "  [%d, %s] = %s\n" state (symbol_to_string sym) action_str
  ) sorted_actions;

  print_endline "\nGOTO Table:";
  let goto_entries = Hashtbl.fold (fun (state, sym) next acc ->
    (state, sym, next) :: acc
  ) parser.goto_table [] in
  let sorted_gotos = List.sort (fun (s1, sym1, _) (s2, sym2, _) ->
    let c = Int.compare s1 s2 in
    if c = 0 then compare_symbol sym1 sym2 else c
  ) goto_entries in
  List.iter (fun (state, sym, next) ->
    Printf.printf "  [%d, %s] = %d\n" state (symbol_to_string sym) next
  ) sorted_gotos

File: rust_version/cfg-parser/src/slr1.rs
//! SLR(1) bottom-up parser implementation.
//!
//! This module implements a shift-reduce SLR(1) parser using LR(0) automaton
//! with lookahead from FOLLOW sets.

use crate::error::{GrammarError, Result};
use crate::first_follow::FollowSets;
use crate::grammar::{Grammar, Production};
use crate::symbol::{string_to_symbols, Symbol};
use std::collections::{HashMap, HashSet, VecDeque};

/// An LR(0) item: a production with a dot position.
///
/// For example: A → α•β is represented as (Production, position)
/// where position is the index of the dot.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
struct Item {
    production: Production,
    dot_position: usize,
}

impl Item {
    fn new(production: Production, dot_position: usize) -> Self {
        Self {
            production,
            dot_position,
        }
    }

    /// Returns the symbol after the dot, if any.
    fn symbol_after_dot(&self) -> Option<Symbol> {
        self.production.rhs.get(self.dot_position).copied()
    }

    /// Checks if the dot is at the end (reduce item).
    fn is_reduce_item(&self) -> bool {
        self.dot_position >= self.production.rhs.len()
    }
}

/// A state in the LR(0) automaton (set of items).
type ItemSet = HashSet<Item>;

/// SLR(1) action.
#[derive(Debug, Clone, PartialEq, Eq)]
enum Action {
    Shift(usize),
    Reduce(Production),
    Accept,
}

/// SLR(1) parser.
#[derive(Debug)]
pub struct SLR1Parser {
    grammar: Grammar,
    augmented_start: Symbol,
    states: Vec<ItemSet>,
    /// ACTION table: (state, terminal/end_marker) → Action
    action_table: HashMap<(usize, Symbol), Action>,
    /// GOTO table: (state, nonterminal) → state
    goto_table: HashMap<(usize, Symbol), usize>,
}

impl SLR1Parser {
    /// Builds an SLR(1) parser from a grammar.
    pub fn build(grammar: Grammar, follow_sets: FollowSets) -> Result<Self> {
        // Create augmented grammar with S' → S
        let start = grammar.start_symbol();
        let augmented_start = Symbol::Nonterminal('\'');
        let start_production = Production::new(augmented_start, vec![start]);

        // Build LR(0) automaton
        let (states, transitions) = Self::build_lr0_automaton(&grammar, &start_production);

        // Build ACTION and GOTO tables
        let (action_table, goto_table) = Self::build_tables(
            &grammar,
            &states,
            &transitions,
            &follow_sets,
            augmented_start,
            &start_production,
        )?;

        Ok(Self {
            grammar,
            augmented_start,
            states,
            action_table,
            goto_table,
        })
    }

    /// Computes the closure of a set of items.
    ///
    /// For each item [A → α•Bβ] where B is nonterminal,
    /// add all items [B → •γ] for each production B → γ.
    fn closure(grammar: &Grammar, items: ItemSet) -> ItemSet {
        let mut result = items;
        let mut changed = true;

        while changed {
            changed = false;
            let current = result.clone();

            for item in &current {
                if let Some(symbol) = item.symbol_after_dot() {
                    if symbol.is_nonterminal() {
                        for production in grammar.get_productions(symbol) {
                            let new_item = Item::new(production.clone(), 0);
                            if !result.contains(&new_item) {
                                result.insert(new_item);
                                changed = true;
                            }
                        }
                    }
                }
            }
        }

        result
    }

    /// Computes goto(I, X) - the set of items obtained by moving dot over X.
    fn goto(grammar: &Grammar, items: &ItemSet, symbol: Symbol) -> ItemSet {
        let mut moved = ItemSet::new();

        for item in items {
            if let Some(sym) = item.symbol_after_dot() {
                if sym == symbol {
                    let new_item = Item::new(item.production.clone(), item.dot_position + 1);
                    moved.insert(new_item);
                }
            }
        }

        Self::closure(grammar, moved)
    }

    /// Builds the canonical LR(0) collection of item sets.
    fn build_lr0_automaton(
        grammar: &Grammar,
        start_production: &Production,
    ) -> (Vec<ItemSet>, HashMap<(usize, Symbol), usize>) {
        let initial_item = Item::new(start_production.clone(), 0);
        let initial_state = Self::closure(grammar, HashSet::from([initial_item]));

        let mut states = vec![initial_state.clone()];
        let mut transitions: HashMap<(usize, Symbol), usize> = HashMap::new();
        let mut worklist: VecDeque<usize> = VecDeque::new();
        worklist.push_back(0);

        while let Some(state_id) = worklist.pop_front() {
            let state = states[state_id].clone();

            // Get all symbols that can be shifted
            let mut symbols = HashSet::new();
            for item in &state {
                if let Some(symbol) = item.symbol_after_dot() {
                    symbols.insert(symbol);
                }
            }

            // For each symbol, compute goto and add new states
            for symbol in symbols {
                let next_state = Self::goto(grammar, &state, symbol);

                if !next_state.is_empty() {
                    // Check if this state already exists
                    if let Some(existing_id) = states.iter().position(|s| s == &next_state) {
                        transitions.insert((state_id, symbol), existing_id);
                    } else {
                        let new_id = states.len();
                        states.push(next_state);
                        worklist.push_back(new_id);
                        transitions.insert((state_id, symbol), new_id);
                    }
                }
            }
        }

        (states, transitions)
    }

    /// Builds ACTION and GOTO tables for SLR(1).
    #[allow(clippy::too_many_arguments)]
    fn build_tables(
        _grammar: &Grammar,
        states: &[ItemSet],
        transitions: &HashMap<(usize, Symbol), usize>,
        follow_sets: &FollowSets,
        augmented_start: Symbol,
        _start_production: &Production,
    ) -> Result<(
        HashMap<(usize, Symbol), Action>,
        HashMap<(usize, Symbol), usize>,
    )> {
        let mut action_table = HashMap::new();
        let mut goto_table = HashMap::new();

        for (state_id, state) in states.iter().enumerate() {
            for item in state {
                if !item.is_reduce_item() {
                    // Shift items: [A → α•aβ] where a is terminal
                    if let Some(symbol) = item.symbol_after_dot() {
                        if symbol.is_terminal() || symbol.is_end_marker() {
                            if let Some(&next_state) = transitions.get(&(state_id, symbol)) {
                                let key = (state_id, symbol);
                                if action_table.contains_key(&key) {
                                    return Err(GrammarError::SLR1ShiftReduceConflict {
                                        state: state_id,
                                        symbol: symbol.to_string(),
                                    });
                                }
                                action_table.insert(key, Action::Shift(next_state));
                            }
                        }
                    }
                } else {
                    // Reduce items: [A → α•]
                    if item.production.lhs == augmented_start {
                        // Accept item: [S' → S•]
                        let key = (state_id, Symbol::EndMarker);
                        action_table.insert(key, Action::Accept);
                    } else {
                        // Reduce on FOLLOW(A)
                        let follow_a = follow_sets
                            .get(&item.production.lhs)
                            .cloned()
                            .unwrap_or_default();

                        for symbol in follow_a {
                            let key = (state_id, symbol);

                            if let Some(existing) = action_table.get(&key) {
                                match existing {
                                    Action::Shift(_) => {
                                        return Err(GrammarError::SLR1ShiftReduceConflict {
                                            state: state_id,
                                            symbol: symbol.to_string(),
                                        });
                                    }
                                    Action::Reduce(other_prod) => {
                                        return Err(GrammarError::SLR1ReduceReduceConflict {
                                            state: state_id,
                                            symbol: symbol.to_string(),
                                            prod1: other_prod.to_string(),
                                            prod2: item.production.to_string(),
                                        });
                                    }
                                    Action::Accept => {}
                                }
                            } else {
                                action_table.insert(key, Action::Reduce(item.production.clone()));
                            }
                        }
                    }
                }
            }

            // Build GOTO table for nonterminals
            for (key, &next_state) in transitions {
                let (src, symbol) = key;
                if *src == state_id && symbol.is_nonterminal() {
                    goto_table.insert((state_id, *symbol), next_state);
                }
            }
        }

        Ok((action_table, goto_table))
    }

    /// Parses an input string using SLR(1) shift-reduce algorithm.
    pub fn parse(&self, input: &str) -> bool {
        // Convert input to symbols and add $
        let mut input_symbols = string_to_symbols(input);
        input_symbols.push(Symbol::EndMarker);

        // Initialize stack with state 0
        let mut stack: Vec<usize> = vec![0];
        let mut symbol_stack: Vec<Symbol> = Vec::new();
        let mut input_index = 0;

        loop {
            if input_index >= input_symbols.len() {
                return false;
            }

            let state = *stack.last().unwrap();
            let current_symbol = input_symbols[input_index];
            let key = (state, current_symbol);

            let action = self.action_table.get(&key);

            match action {
                Some(Action::Accept) => return true,
                Some(Action::Shift(next_state)) => {
                    // Push symbol and next state
                    stack.push(*next_state);
                    symbol_stack.push(current_symbol);
                    input_index += 1;
                }
                Some(Action::Reduce(production)) => {
                    // Pop |rhs| symbols and states
                    let rhs_len = if production.rhs == vec![Symbol::Epsilon] {
                        0
                    } else {
                        production.rhs.len()
                    };

                    for _ in 0..rhs_len {
                        stack.pop();
                        symbol_stack.pop();
                    }

                    // Get state at top of stack after popping
                    let state_after_pop = *stack.last().unwrap();

                    // Find next state via GOTO
                    let goto_key = (state_after_pop, production.lhs);
                    if let Some(&next_state) = self.goto_table.get(&goto_key) {
                        stack.push(next_state);
                        symbol_stack.push(production.lhs);
                    } else {
                        return false;
                    }
                }
                None => return false,
            }
        }
    }
}
